### PR TITLE
HW-step2 CRUD functionality to the website

### PR DESCRIPTION
**Pull request by ss77995ss**
# Description
- Add CRUD functionality that user can manipulate table's data
- Create -> Use a form to add new row for the table, seq cannot be duplicated
- Read -> Find rows that contain input data by seq or by owner
- Update -> Modify selected row's data, seq cannot be duplicated
- Delete -> Delete selected row's data
# 10/4 Update
- Change form to controlled form
- Apply airbnb eslint rules
- Modify file structure
# 10/8 Update
- When remove `eslint:recommended` in `eslint.json` a lot of errors popped up
- Fix these errors and adjust codes that comments mentioned
# Encounter Problems
- Cannot find a more efficient way to deliver form's data
- Have to use a lot of state to store data

### CODE REVIEW

File: .eslintrc.json [hunk 1]
```diff
@@ -0,0 +1,55 @@
+{
+    "env": {
+        "browser": true,
+        "commonjs": true,
+        "es6": true
+    },
+    "extends": [
```
Comment (chunming-c) [inline .eslintrc.json]: I highly recommended to lint our javascript according to [Airbnb JavaScript Style Guide](https://github.com/airbnb/javascript), which is mostly reasonable and universla approach to JavaScript. 

`npm install eslint-config-airbnb --save-dev`

```diff
{
  "extends": [
+    "airbnb",
    "eslint:recommended",
    "plugin:react/recommended",
    "plugin:jest/recommended"
  ],
}
```

File: package.json [hunk 2]
```diff
@@ -29,6 +29,9 @@
     "enzyme": "^3.6.0",
     "enzyme-adapter-react-16": "^1.5.0",
     "enzyme-to-json": "^3.3.4",
+    "eslint": "^5.6.0",
+    "eslint-plugin-jest": "^21.22.1",
+    "eslint-plugin-react": "^7.11.1",
     "identity-obj-proxy": "^3.0.0",
```
Comment (chunming-c) [inline package.json:41]: It seems that we don't have `prop-types` in our `dependencies` list.

File: src/client/Table.js [hunk 3]
```diff
@@ -17,6 +18,9 @@ class Table extends React.Component{
           title={item.title}
           owner={item.owner}
           priority={item.priority}
```
Comment (chunming-c) [inline src/client/Table.js]: Repetition is bad smell for coding. Here I suggest to destruct your `item` and `props` here. It would highly improves readability and cleanness for code base.

ref: [javascript#destructuring](https://github.com/airbnb/javascript#destructuring)

```js
const { seq, status, category, title, owner } = item;

 return(
  <TableRow
    seq={seq}
    status={status}
  />
);
```

Comment (chunming-c) [reply src/client/Table.js]: One thing we need to think it further, should we disassemble our `row` and pass down as `props` here?
Pass `row` directly might be more flexible if we want to do some modification in `row` or `cell` level.

in `Table.js`.
```diff
<TableRow
- status={item.status}
- category={item.category}
- title={item.title}
- owner={item.owner}
- ...
+ row={row}
/>
```

in `TableRow.js`
```js
const TableRow = (props) => {
  const { seq, status, ... } = props.row; // destruct your row here.
}
```

File: src/client/TableRow.js [hunk 4]
```diff
@@ -1,21 +1,32 @@
 import React from 'react';
 import PropTypes from 'prop-types';
+import Delete from './Delete';
+import Update from './Update';
 
 class TableRow extends React.Component{
```
Comment (chunming-c) [inline src/client/TableRow.js]: If the responsibility of the component is rendering only, I suggest to make it `functional`. It becomes concise and also indicate that no `lifecycle` method inside this component, which means UI only.

ref: [functional-and-class-components](https://reactjs.org/docs/components-and-props.html#functional-and-class-components)

```js
const TableRow = (props) => {
  return { // your DOM here.
    ...
  }
}
```

File: src/client/CreateRow.js [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+  handleSubmit(e){
+    e.preventDefault();
+    this.props.onSubmit(e);
+  }
+
+  render(){
+    const reveal = this.props.isClicked ? 'block' : 'none';
+    return(
+      <div className="create__row" style={{display: reveal}}>
+        <form className="create__form" onSubmit={this.handleSubmit}>
```
Comment (chunming-c) [inline src/client/CreateRow.js]: Instead classical html form submit,  make this component controllable would be very helpful for data maintenance.

ref: [controlled-components](https://reactjs.org/docs/forms.html#controlled-components)


```js
const React, { Component } from 'react'

class CreateRow extends Component {
  constructor(props) {
    super(props);
    this.state = {
      // initial state of your form, e.g.
     status: '',
     ... 
    }

    render () {
      <form className="create__form" onSubmit={this.handleSubmit}>
        <input type="text" value={this.state.status}  onChange={handleChange} />
        ...
      </form>
    }
  }

}

```

**Comment by chunming-c** [comment]
According to the problem we met, there are **3** parts we could make it better, as our application getting complicated and complicated.

* `controlled-component`: 2-way-data-binding allows to manipulate data easier.
* `folder classification`:  As our application getting bigger and bigger, classify our component into folders would be helpful to organize.
* `flux data flow`: A central data store.

File: src/client/site/tableRow.js [hunk 6]
```diff
@@ -0,0 +1,48 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import { rowPropType, updatePropType } from '../data/shape';
+import Delete from '../action/delete';
+import Update from '../action/update';
+
+function TableRow (props){
+  const { index, row, updatedStyle, onDelete, onUpdate } = props;
```
Comment (chunming-c) [inline src/client/site/tableRow.js]: I suggest to destruct `row` here because its attributes were used frequently.

```js
// Destruct respectively is good
const { index, row, updatedStyle, onDelete, onUpdate } = props;
const { seq, status, category ... } = row

// nested destruction also available
const {
  index,
  row: {
    seq,
    status,
    category,
    title,
    owner,
    priority,
  },
  updatedStyle,
  onDelete,
  onUpdate
} = props;

```

File: src/client/site/tableRow.js [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+  const isEven = (index % 2 === 0);
+  const className = isEven ? 'table__row table__row--even' : 'table__row table__row--odd';
+  const rowStyle = (updatedStyle.seq === row.seq) ? updatedStyle.color : '';
```
Comment (chunming-c) [inline src/client/site/tableRow.js]: `rowStyle` seems to be `object` by its naming. Instead, I recommended to make some adjustment like following.

```js
  rowStyle = {
    backgroundColor: updatedStyle.seq === row.seq ? updatedStyle.color : ''
       // The benefit is that it keep flexible if we want to add inline style in the future.
  };

  return (
    <tr className={className} style={rowStyle}> { ... }</tr>
```

File: src/client/site/tableRow.js [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+  const target = {
+    seq: row.seq
+  };
+
+  return (
+    <tr className={className} style={{backgroundColor: rowStyle}}>
+      <td>{row.seq}</td>
```
Comment (chunming-c) [inline src/client/site/tableRow.js]: After destructuring, it becomes more concise. 
```js
return (
  <tr>
    <td>{seq}</td>
    <td>{status}</td>
    <td>{category}</td>
    <td>{title}</td>
    <td>{owner}</td>
    <td>{priority}</td>
  </tr>
);
```

File: src/client/site/tableRow.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+      <td>{row.status}</td>
+      <td>{row.category}</td>
+      <td>{row.title}</td>
+      <td>{row.owner}</td>
+      <td>{row.priority}</td>
+      <td>
+        <Delete target={target} onDelete={onDelete} />
```
Comment (chunming-c) [inline src/client/site/tableRow.js]: If target only contains `seq` (sequence), I suggest to flatten and not to pass `object`. Precisely, `seq` is enough.

```js
<Delete  seq={seq} onUpdate={onUpdate} />
<Update seq={seq} onUpdate={onUpdate} />

// or you could rename your prop if you like
<Update rowKey={seq} onUpdate={onUpdate} />
```

File: src/client/site/tableRow.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+        <Update target={target} onUpdate={onUpdate} />
+      </td>
+    </tr>
+  );
+}
+
+TableRow.propTypes = {
+  index: PropTypes.number,
+  row: rowPropType,
```
Comment (chunming-c) [inline src/client/site/tableRow.js]: For `<TableRow />`, `row` seems mandatory and essential.
Mark as `required` looks reasonable.

```js
TableRow.propTypes = {
  ...
  row: rowPropType.isRequired,
};
``

Comment (chunming-c) [reply src/client/site/tableRow.js]: Expect `row`, `onDelete` and `onUpdate` looks indispensable too.

File: src/client/data/shape.js [hunk 11]
```diff
@@ -0,0 +1,15 @@
+import PropTypes from 'prop-types';
```
Comment (chunming-c) [inline src/client/data/shape.js]: `shape` seems irrelevant to `data`. Create a new folder name `type` or `shape` would be suitable.

```diff
  ├── client
  │   ├── App.js
  │   ├── action
  │   ├── assets
+ │   ├── types
  │   ├── data
  │   ├── index.js
  │   └── site
```

File: src/client/action/createRow.js [hunk 12]
```diff
@@ -0,0 +1,94 @@
...
+    e.preventDefault();
+    onSubmit(formData);
+  }
+
+  render(){
+    const { seq, status, category, title, owner, priority } = this.state;
+    const { isClicked } = this.props;
+    const reveal = isClicked ? 'block' : 'none';
+    return(
+      <div className="create__row" style={{display: reveal}}>
```
Comment (chunming-c) [inline src/client/action/createRow.js]: Bad smell for high duplication. Funtion-nize for your `input` and `label` for `form` to improve readability.

```js
  const labelNameList = ['seq', 'status', 'category', 'title', 'owner', 'priority' ];

    return(
      <div className="create__row" style={{display: reveal}}>
        <form className="create__form" onSubmit={this.handleSubmit}>
          {
            labelNameList.map((index, name) => {
              return (
                <label htmlFor={name} key={`create-row-${name}`}>
                  {name}
                  <input
                    type="text"
                    name={name}
                    value={this.state[name]}
                    onChange={this.handleInputChange}
                    required
                  />
                </label>
              );
            })
          }
          <input type="submit" text="Submit" />
        </form>
      </div>
```

Comment (chunming-c) [reply src/client/action/createRow.js]: Also, `seq` should be taken as kind of row `key` and I thought it could be editable.

File: src/client/site/home.js [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+import CreateRow from '../action/createRow';
+import Read from '../action/read';
+import Table from './table';
+import Footer from './footer';
+
+class Home extends React.Component {
+  constructor(props){
+    super(props);
+    this.state = {
+      data: data,
```
Comment (chunming-c) [inline src/client/site/home.js]: What's the difference between `data` and `root`?

File: src/client/site/table.js [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+import { rowPropType, updatePropType } from '../data/shape';
+import TableRow from './tableRow';
+
+function Table (props){
+  const { tableData, receiveDeleteTarget, receiveUpdateTarget, updatedStyle } = props;
+  let tableRow = tableData.map((item, index) => {
+    return(
+      <TableRow
+        key={item.seq}
+        index={index+1}
```
Comment (chunming-c) [inline src/client/site/table.js]: Space should be placed between `operator`.

```js
index={index + 1}
```

File: .eslintrc.json [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+        "airbnb",
+        "eslint:recommended",
```
Comment (chunming-c) [inline .eslintrc.json]: I suggest not to include `eslint:recommended` which seems to overwrite lots of rules we used to adopting.

Comment (chunming-c) [reply .eslintrc.json]: Additionally, inappropriate `indentation` was found in this file. 
Not `4 spaces`, `2 spaces` instead.

```json
{
  "env": {
    ...
  }
}
```

**Review by chunming-c — COMMENTED** [review]
Well done but there are some advices for improvement. Please kindly have a check. 👍

File: src/client/site/home.js (hunk 13 above)
Comment (ss77995ss) [reply src/client/site/home.js]: I want to use `root` to save the origin table because I use `data` to create a table every time when it finishes an action and some action like `read` while affect the table's length. 

Ex.: If I only get one row after `read` action activate, then my table data will have only one element.

File: src/client/index.js [hunk 16]
```diff
@@ -1,6 +1,6 @@
-import App from './app';
 import React from 'react';
 import { render } from 'react-dom';
+import App from './app';
```
Comment (chunming-c) [inline src/client/index.js:3]: File typo here.

```diff
  ├── client
- │   ├── App.js
+ │   ├── app.js
  │   ├──  ... 
  │   └── types
```

File: src/client/site/home.js [hunk 17]
```diff
@@ -0,0 +1,170 @@
+import React from 'react';
+import '../assets/home.css';
+import fakedata from '../data/data.json';
```
Comment (chunming-c) [inline src/client/site/home.js]: It should be `camelCase`, e.g. `fakeData`.

File: src/client/action/createRow.js [hunk 18]
```diff
@@ -0,0 +1,97 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+
+class CreateRow extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = {
```
Comment (chunming-c) [inline src/client/action/createRow.js:7]: It seems to mismatching for type of `seq`. Expected `number` but got `string` if we create or update the row.

```js
export const rowPropType = PropTypes.shape({
  seq: PropTypes.number,
  status: PropTypes.string,
  category: PropTypes.string,
  title: PropTypes.string,
  owner: PropTypes.string,
  priority: PropTypes.string,
```

File: package.json [hunk 19]
```diff
@@ -29,6 +30,12 @@
     "enzyme": "^3.6.0",
     "enzyme-adapter-react-16": "^1.5.0",
     "enzyme-to-json": "^3.3.4",
+    "eslint": "^5.6.0",
```
Comment (chunming-c) [inline package.json:35]: Once we've adopted `eslint` for linting, please kindly create a `script` to allow others run your lint, not just by editor plugins. e.g.

```json
"scripts": {
  ...
  "lint": "./node_modules/.bin/eslint src",
  "lint:fix": "./node_modules/.bin/eslint src --fix"
},
```


ref: https://eslint.org/docs/user-guide/command-line-interface#options

Comment (chunming-c) [reply package.json:35]: By doing so, it allowed to lint your code not only `local` but also `CI` environment.
Please feel free to enhance your automatic test workflow.

File: test/action/createRow.test.js [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+  describe('WHEN input changed', () => {
+    it('should change state value', () => {
+      wrapper.instance().handleInputChange(mockInput);
+      expect(wrapper.state('seq')).toBe(123);
+    });
+  });
+
+  describe('WHEN submit is clicked', () => {
+    it('should call handleSubmit', () => {
+      wrapper.find('form').simulate('submit', { preventDefault() {} });
```
Comment (chunming-c) [inline test/action/createRow.test.js]: You may need a realistic HTML click `event`, not just mimicking.
Besides, for `jest`, `jsdom` was picked up as default as its `testEnvironment`.

```js
describe('when submit button is clicked', () => {
  it('should call handleSubmit', () => {
    wrapper.find('form').simulate('submit', new Event('click'));
    expect(mockFunc).toHaveBeenCalled();
  });
});
```
https://github.com/jsdom/jsdom
https://jestjs.io/docs/en/configuration.html#testenvironment-string

**Review by chunming-c — APPROVED** [review]
LGTM
